### ISSUE TITLE
Invoke-PnPSiteTemplate throws Access denied error on SiteSettings

### ISSUE THREAD

**Issue by klaaslauwers** [pnp/pnpframework#357]
**Expected behavior**
Execute without issue

**Actual behavior**
Invokes the site template correctly, except for site settings. The quick navigation is not provisioned successfully.

If I add the following switch to `Get-PnPSiteTemplate` and `Invoke-PnPSiteTemplate`, the site template succeeds:
`-ExcludeHandlers SiteSettings`. I need to add the quick launch manually afterwards. 

**Steps to reproduce behavior**
```
$connection = Connect-PnPOnline -Url $url -Credentials $credPnP -ReturnConnection -ErrorAction Stop 
$siteConnection = Connect-PnPOnline -Url $siteUrl -Credentials $credPnP -ReturnConnection -ErrorAction Stop 
Get-PnPSiteTemplate -Connection $connection -Out $templateXmlName -IncludeAllPages -PersistBrandingFiles -Force
Invoke-PnPSiteTemplate -Connection $siteConnection -Path $templateXmlName -IgnoreDuplicateDataRowErrors -OverwriteSystemPropertyBagValues -ClearNavigation 
```
`Invoke-PnPSiteTemplate`: Access denied. You do not have permission to perform this action or access this resource. Full exception below.

**What is the version of the Cmdlet module you are running?**
PnP.PowerShell version 1.5.0

**Which operating system/environment are you running PnP PowerShell on?**
Errors are received locally as well as in Azure RunBook / Azure Function

**Exception**
In Dutch, apologies, but it says "Access Denied". It looks like it's on the check for feature activation; nevertheless, the features are properly activated.
```
Message          : Toegang geweigerd. U bent niet gemachtigd om deze bewerking uit te voeren of deze bron te openen.
Stacktrace       :    bij Microsoft.SharePoint.Client.ClientRequest.ProcessResponseStream(Stream responseStream)
                      bij Microsoft.SharePoint.Client.ClientRequest.ProcessResponse()
                      bij Microsoft.SharePoint.Client.ClientRequest.<ExecuteQueryToServerAsync>d__53.MoveNext()
                   --- Einde van stacktracering vanaf vorige locatie waar uitzondering is opgetreden ---
                      bij System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
                      bij System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task
                   )
                      bij Microsoft.SharePoint.Client.ClientRequest.<ExecuteQueryAsync>d__39.MoveNext()
                   --- Einde van stacktracering vanaf vorige locatie waar uitzondering is opgetreden ---
                      bij System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
                      bij System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task
                   )
                      bij Microsoft.SharePoint.Client.ClientRuntimeContext.<ExecuteQueryAsync>d__65.MoveNext()
                   --- Einde van stacktracering vanaf vorige locatie waar uitzondering is opgetreden ---
                      bij System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
                      bij System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task
                   )
                      bij Microsoft.SharePoint.Client.ClientContext.<ExecuteQueryAsync>d__28.MoveNext()
                   --- Einde van stacktracering vanaf vorige locatie waar uitzondering is opgetreden ---
                      bij System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
                      bij Microsoft.SharePoint.Client.ClientContextExtensions.<ExecuteQueryImplementation>d__6.MoveNext
                   ()
                   --- Einde van stacktracering vanaf vorige locatie waar uitzondering is opgetreden ---
                      bij System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
                      bij System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task
                   )
                      bij Microsoft.SharePoint.Client.FeatureExtensions.<IsFeatureActiveInternal>d__12.MoveNext()
                   --- Einde van stacktracering vanaf vorige locatie waar uitzondering is opgetreden ---
                      bij System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
                      bij System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task
                   )
                      bij Microsoft.SharePoint.Client.FeatureExtensions.IsFeatureActive(Web web, Guid featureID)
                      bij Microsoft.SharePoint.Client.SiteExtensions.IsCommunicationSite(Site site)
                      bij PnP.Framework.Provisioning.ObjectHandlers.ObjectSiteSettings.ProvisionObjects(Web web, Provis
                   ioningTemplate template, TokenParser parser, ProvisioningTemplateApplyingInformation applyingInforma
                   tion)
                      bij PnP.Framework.Provisioning.ObjectHandlers.SiteToTemplateConversion.ApplyRemoteTemplate(Web we
                   b, ProvisioningTemplate template, ProvisioningTemplateApplyingInformation provisioningInfo, Boolean
                   calledFromHierarchy, TokenParser tokenParser)
                      bij PnP.PowerShell.Commands.Provisioning.Site.InvokeSiteTemplate.ExecuteCmdlet()
                      bij PnP.PowerShell.Commands.PnPSharePointCmdlet.ProcessRecord()
```
Linked issue: https://github.com/pnp/pnpframework/issues/193

**Comment by jackpoz** [comment]
does this happen with a newly created source site too as template ? I created a new site as source, a new site as destination, ran your commands (fixing Invoke-PnPSiteTemplate to pass  -Connection $siteConnection instead of $siteUrl) and I didn't get any error

**Comment by klaaslauwers** [comment]
@jackpoz Thank you for the reply!

For a newly created site it works as expected. At the moment I start modifying the site settings, things go wrong. 

Things I've tested: 


1. Blank site, no changes: OK
2. 1 + Reporting, Document Sets and Document ID feature enabled + language settings (removed all translations): **Access Denied**, but everything provisioned correctly 
3. 2 + document library "Dossier" and new site page: **Access Denied**, but everything correctly provisioned
4. 3 + some folders added in Dossier and `Add-PnPListFoldersToSiteTemplate -Connection $connection -List "Dossier" -Path $templateXmlName -Recursive`: **Access Denied**, but everything provisioned correctly 
5. 4 + custom quick launch: **Access Denied, navigation not provisioned!**

I've added the full template XML (so step 5) for your reference (I've modified the domain name + users though).
[Template.zip](https://github.com/pnp/pnpframework/files/6574317/Template.zip)

P.S. Oops, typo in the site connection indeed.

**Comment by jackpoz** [comment]
I tried applying the template to an English site and while testing it I received some Access Denied errors when trying to update the Site Columns that were created by a previous provisioning
![image](https://user-images.githubusercontent.com/1153754/120373670-e7c4d500-c318-11eb-8b43-7fbe201297b8.png)

Your callstack looks different as it contains `Microsoft.SharePoint.Client.SiteExtensions.IsCommunicationSite(Site site)` instead.

After I deleted the Site Columns and tried again I didn't receive any access denied anymore, just a warning about -ClearNavigation and the value stored in the template:
![image](https://user-images.githubusercontent.com/1153754/120373894-307c8e00-c319-11eb-8563-f1dfe153dbf1.png)

This is the navigation of the site:
![image](https://user-images.githubusercontent.com/1153754/120373970-3e321380-c319-11eb-809d-d2b828c25006.png)

I tried again a few times with English site and had the same behavior of Access Denied when updating some Site Columns. I tried then with a Dutch site and that time I didn't receive any error while creating the Site Columns or lists but at the end I received Access Denied with the same callstack as you:
![image](https://user-images.githubusercontent.com/1153754/120375573-3f644000-c31b-11eb-89cb-58315ad795de.png)

The exception points to line 77 of ObjectSiteSettings.cs
https://github.com/pnp/pnpframework/blob/e40396c4b3ef86d974ed39f0fd6426599a652cda/src/lib/PnP.Framework/Provisioning/ObjectHandlers/ObjectSiteSettings.cs#L66-L77

If you check the code above, you can see that the settings of the site are updated but not submitted yet, as `isDirty = true` delays the update at the end of the function. On the other hand, `site.IsCommunicationSite()` triggers an ExecuteQuery() which submits the updated site settings and triggers the exception.

If I comment the `<pnp:SiteSettings>` element in your template and apply it to a Dutch site, I don't get any access denied anymore.

It seems like the settings get sent with the wrong context, as the whole "isDirty" pattern doesn't really work in this case as there are other functions that trigger an ExecuteQuery anyway. If I call ExecuteQuery right away then it works without Access Denied errors.

I created a pull request at https://github.com/pnp/pnpframework/pull/360 that ensures ExecuteQuery is called on the right Context. I didn't receive any Access Denied error with that PR.

**Comment by jackpoz** [comment]
@klaaslauwers as a workaround you could remove the <pnp:SiteSettings> section from the template

**Comment by pieronegri** [comment]
Had the same issue, for me it worked 
Invoke-PnPSiteTemplate -Path template.xml -ExcludeHandlers SiteSecurity

**Comment by klaaslauwers** [comment]
Hi @pieronegri,

I don't want to exclude handlers; the site security is one of the things that is really needed :-). I'm forced to schedule the script in a cronjob on a Unix server; awaiting the fix in Azure RunBooks.

**Comment by rmaersch** [comment]
This thread is old. Tell me if I need to make a new one but I have a related issue. By grabbing the source for PnP.Powershell and PnP.Framework I have traced an issue with the navigation being build using a template. I was using PnP.Office.Core in the past and it worked fine. I then made the switch to see if it fixed the problem but it did not. Not sure if this is server side or not. I found a line in Framework that calls web.AllProperties.ClearObjectData() which I get the access denied.  I question the idea of clearing all the properties of the web object to start with. It seems pretty all encompassing of a command before an update to navigation nodes. What it is checking for is if "__PublishingFeatureActivated". 
I may play with the code and remove the offending line seeing it is trying to clear all the properties and then see if a value exists.  Line 874 of WebExtensions.cs
